syntax = "proto3";

message Product {
    int32   id              = 1;
    int32   company         = 2;
    string  name            = 3;
}

message Code {
    string  message         = 1;
    int32   company         = 2;
    string  created_at      = 3;
    string  activated_at    = 4;
    Product product         = 5;
}

// file should be embedded in another object, like a certification
message HostedFile {
    string name             = 1;
    string md5_hash         = 2;
    int32  size             = 3;
    string url              = 4;
}

message Certification {
    string key              = 1;
    string name             = 2;
    string certifier_url    = 3;
    string instructions     = 4; // instructions on how to look up the document at the url
    string valid_from       = 5;
    string valid_to         = 6;

    HostedFile file         = 7;
}

message Farm {
    string key              = 1;
    string name             = 2;
    repeated
        string address      = 3;
    Location location       = 4;
    repeated string
        certifications      = 5;
}

message Location {
    float       lat         = 1;
    float       lng         = 2;
    string      description = 3;
}

message Harvest {
    string      key         = 1;
    string      country     = 2;
    int32       year        = 3;
    int32       month       = 4;
    Location    location    = 5;
    repeated
        string  farms       = 6;
    repeated
        string  shipments   = 7;
}

message Shipment {
    string      key         = 1;
    int32       kg          = 2;  // volume in KG
    string      ship_name   = 3;
    Location    source      = 4;
    Location    destination = 5;
    string      shipped_at  = 6;
    string      recieved_at = 7;
    repeated
        string  extra_info  = 8;  // list of document data, as strings
}

message Roast {
    string      key         = 1;
    string      roasted_at  = 2;
    Location    location    = 3;
    repeated
        string  harvests    = 4;
}

message Events {
    message MintCode {
        string message      = 1;
        int32  company      = 2;
        string created_at   = 3;
    }

    message ActivateCode {
        string message = 1;
        string activated_at = 2;
    }

    message FarmAddCert { string key = 1; }

    // generic add-related-field-to-object event, so we don't have to
    // spawn a million of them.  Object update events should be fine.
    // it is up to the handler to process these properlyp
    message AddRelated {
        string action = 1;
        string object_key = 2;
        string related_key = 3;
    }
}

message CoffeeChainEvents {
    oneof payload {
        Events.MintCode     mint_code       = 1;
        Events.ActivateCode activate_code   = 2;
        Certification       cert_create     = 3;
        Harvest             harvest_create  = 4;
        Farm                farm_create     = 5;
        Events.FarmAddCert  farm_add_cert   = 6;
        Events.AddRelated   add_related     = 7;
        Shipment            shipment_create = 8;
        Roast               roast_create    = 9;
    }
}